mlir/lib/Dialect/Rock/IR/RockDialect.cpp

```diff
@@ -1673,6 +1673,91 @@ LogicalResult ReduceOp::verify() {
   return success();
 }
 
+//===-----------------------------------------------------===//
+// Blockwise_ReduceOp
+//===-----------------------------------------------------===//
+
+LogicalResult BlockwiseReduceOp::verify() {
+  ArrayAttr inputViewArrayAttr = getInputRegViewAttr();
+  // This view should be {bid, tid, iter} to {bid, d0, ... , Dr , ... , dn};
+  // Moreover, {bid, d0, ... , Dr , ... , dn} --> {D0, ... , Dr , ... , Dn}
+  // should not be a part of this view where the latter is the larger
+  // input tensors that being reduced accross blocks.
+
+  TransformMapAttr inputView = inputViewArrayAttr[0].cast<TransformMapAttr>();
+  ArrayRef<int64_t> inputTensorShape = inputView.getLowerBounds().asArrayRef();
+  ArrayRef<int64_t> inputThreadView = inputView.getUpperBounds().asArrayRef();
+  ArrayRef<int64_t> wsShape = getWorkspaceBuffer().getType().getShape();
+  int64_t blockSize = getBlockSize();
+  int64_t gridSize = getGridSize();
+
+  gpu::AddressSpaceAttr inMemSpaceAttr =
+      getInput()
+          .getType()
+          .getMemorySpace()
+          .dyn_cast_or_null<gpu::AddressSpaceAttr>();
+  if (!inMemSpaceAttr) {
+    return emitError("No gpu memspace attr found in input memref; the input "
+                     "memref should be in regs");
+  } else {
+    if (inMemSpaceAttr.getValue() != gpu::AddressSpace::Private) {
+      return emitError("input should be in regs.");
+    }
+  }
+
+  gpu::AddressSpaceAttr outMemSpaceAttr =
+      getOutput()
+          .getType()
+          .getMemorySpace()
+          .dyn_cast_or_null<gpu::AddressSpaceAttr>();
+  if (!outMemSpaceAttr) {
+    return emitError("No gpu memspace attr found in output memref; the output "
+                     "memref should be in regs");
+  } else {
+    if (outMemSpaceAttr.getValue() != gpu::AddressSpace::Private) {
+      return emitError("output should be in regs.");
+    }
+  }
+
+  gpu::AddressSpaceAttr wsMemSpaceAttr =
+      getWorkspaceBuffer()
+          .getType()
+          .getMemorySpace()
+          .dyn_cast_or_null<gpu::AddressSpaceAttr>();
+  if (!wsMemSpaceAttr) {
+    return emitError("No gpu memspace attr found in workspace memref; the "
+                     "workspace memref should be in LDS");
+  } else {
+    if (wsMemSpaceAttr.getValue() != gpu::AddressSpace::Workgroup) {
+      return emitError("workspace should be in LDS.");
+    }
+  }
+
+  if (inputTensorShape[0] != gridSize) {
```

Comment:
Ah, now I get it - coordinate transforms don't have symbols so we can't slice off the workgroup ID component